Python/chapter02/2.6 - Palindrome/miguel_2.6_soln.py

```diff
@@ -0,0 +1,218 @@
+        new_head.next = self.head
+        if self.head is None:
+            # if list is empty and we add
+            # out first element, head AND tail
+            # must point to same node
+            self.tail = new_head
+        self.head = new_head
+        self.size += 1
+
+    def get_node_at(self, index: int) -> Node:
+        if index < 0 or index >= self.size:
+            raise IndexError('list index out of range')
+        n = self.head
+        for i in range(self.size):
+            if i == index:
+                return n
+            n = n.next
+
+    def pop_head(self) -> Node:
+        if self.head is None:
+            raise IndexError('no head to pop')
+        h = self.head
+        h.next = None
+        self.head = self.head.next
+        self.size -= 1
+        return h
+
+    def append(self, ll: 'LinkedList') -> None:
+        self.tail.next = ll.head
+        self.size += ll.size
+        ll.head = None
+        ll.size = 0
+
+    def reverse(self) -> None:
+        """
+        Reverses this linked list in place
+        :return:
+        """
+        if self.head is None:
+            return
+        prev = self.head
+        self.tail = prev
+        curr = prev.next
+        self.tail.next = None
+        while curr is not None:
+            old_next = curr.next
+            curr.next = prev
+            prev = curr
+            curr = old_next
+        self.head = prev
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        if self.head is None:
+            return '<empty>'
+        ll = []
+        n = self.head
+        while n.next is not None:
+            ll.append('{} -> '.format(n.data))
+            n = n.next
+        ll.append(str(n.data))
+        return ''.join(ll)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, LinkedList):
+            return NotImplemented
+        a = self.head
+        b = other.head
+        while a is not None and b is not None:
+            if a.data != b.data:
+                return False
+            # otherwise, advance both pointers
+            a = a.next
+            b = b.next
+        return a is None and b is None
+
+
+def reverse_linked_list(ll: LinkedList) -> LinkedList:
+    """
+    Takes in a linked list and returns a reversed copy
+    :param ll: input linked list
+    :return: reversed linked list
+    """
+    output_ll = LinkedList()
+    n = ll.head
+    while n is not None:
+        output_ll.append_to_head(n.data)
+        n = n.next
+    return output_ll
+
+
+def palindrome(ll: LinkedList) -> bool:
+    """
+    Given a linked list, this function will check if the
+    linked list is a palindrome.
+    A palindrome is a word or phrase that is the same
+    forwards and backwards.
+    Since I will use integer nodes for the linked list,
+    we will be checking if the sequence of numbers
+    in a linked list is a palindrome.
+    Runtime:  O(n)
+    Space Complexity:  O(n)
```

Comment:
is this optimal?

Reply:
I think for runtime yes, but space complexity can be constant if I compare each element in the linked list with its complement (length - I - 1) up until the halfway point without creating a new reversed linked list.

Reply:
yes, runtime you cannot do better than O(n) since you need to at least inspect every single element. but space can be improved as you say.

Reply:
ok, going to write a new function for that.